src/profiler/custom_op_profiler.h

```diff
@@ -0,0 +1,93 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+#ifndef MXNET_CUSTOM_OP_PROFILER_H_
+#define MXNET_CUSTOM_OP_PROFILER_H_
+
+#include <string>
+#include <unordered_set>
+#include <unordered_map>
+#include <thread>
+#include "./profiler.h"
+
+namespace mxnet {
+namespace profiler {
+
+using Tid = std::thread::id;
+using TaskPtr = std::shared_ptr<ProfileTask>;
+
+class CustomOpProfiler {
```

Comment:
make constructor as private to make the class a Singleton

Reply:
protected would be better usually in case you want to extend.

Reply:
Why is this resolved, I still did not see a protected/private constructor?

Reply:
In an earlier version I used protected constructor, but I revert back because make_shared only allows public constructor.

Reply:
It seems use a shared_ptr for singleton class itself is not a good design practice:

Please refer to:
https://stackoverflow.com/questions/55555451/how-to-use-stdmake-shared-with-a-singleton-instance
https://codereview.stackexchange.com/questions/13059/thread-safe-singleton-class-using-stdshared-ptr-in-c11

Why do we need make_shared on a singleton object?

Reply:
Sorry there was a typo in my last comment. I used unique_ptr, but the idea is the same.
I used unique_ptr because CustomOperator used it. Also, because of thread-safety consideration, I used the double-checked locking https://www.baeldung.com/java-singleton-double-checked-locking.

Reply:
This is a good idiom, we should apply to other places.